doc: create notation threat model (#242)
notaryproject/notaryproject: threatmodels/notation-theatmodel.md

| Tampered files               | Tampering              | Mitigated     | High     | File System     | Files read by Notation are tampered, this may lead to failure of notation sign operation | Use trusted infrastructure with right file permissions |
| Disclosure of files          | Information disclosure | Mitigated     | High     | File System     | Files are accessed by attackers, this may lead to disclosure of notation config files and credential files | Use trusted infrastructure with right file permissions. Users can use credential store to securely store the credentials. Users can use notation plugin to access signing keys via remote KMS or a HSM device |
| Malicious plugin             | Tampering              | Mitigated     | High     | Notation Plugin | A malicious plugin is installed, this may lead to arbitrary code being executed by attackers | Verify plugin package before installation or plugin is installed by authorized users |

### Notation verify artifacts against signatures

> ![image](../media/notation-verify.png)

Reviewer: missing image

> ![image](../media/notation-verify.png)

Reviewer 2: For the verification threat model image, since we will support offline verification, we should also add `Read signatures (from File system)` -> `Notation`. During verification, we still need the target artifact descriptor, thus, we should add `Read descriptor (from File system)` -> `Notation` and `Fetch descriptor of artifact (from remote registry)` -> `Notation` as well.

### Threats and Mitigation

|   Title                      | STRIDE threat type     | Threat Status | Priority | Entity          | Description   | Mitigation   |
| ---------------------------- | ---------------------- | ------------- | -------- | --------------- | ------------- | ------------ |
| Compromised signing key      | Information disclosure | Mitigated     | High     | Signer          | Signing keys are compromised, and this may lead to arbitrary artifacts being signed by attackers | Rotate the signing key, and revoke the associated certificates |
| Inaccessible Registry        | Denial of Service      | Mitigated     | High     | Registry        | Registry is not able to service incoming requests or perform up to spec, thus users are unable to publish any artifacts | Users sign offline artifacts and publish via other medias |

> | Serving tampered artifacts   | Tampering              | Mitigated     | High     | Registry        | Data stored in registries are tampered, this may lead to malicious artifacts being signed | Users sign offline artifacts, and store signatures locally or in the registry |

Reviewer 3: should this be marked as mitigated if offline signing/verification is not supported yet (for this issue: https://github.com/notaryproject/roadmap/issues/15)?

Author: Offline signing is planned in RC-4 release, and this threat model is targeting on features delivered in Notation v1.0 GA release.

# Notation Threat Model

## Overview

> Notation aims for signing and validating software artifacts, ensuring they have not been tampered with and providing security policies to determine which validated artifacts are allowed to be used in your systems.

Reviewer 4: Notation is a unified tool for generation and verification of signatures associated with the software artifacts, thus ensuring integrity, authenticity by providing fine grained policies.

Notation is a comprehensive tool for generating and verifying signatures associated with software artifacts. Its purpose is to validate the integrity and authenticity, accomplishing this through the implementation of fine-grained policies.

Author: I will update to the 2nd option.

The following diagram illustrates the architecture and components.

> ![image](../media/notation-system.svg)

Reviewer 5: Please write meaningful `alt` for images for accessibility considerations.

> - **User**: The User could be a signer or a verifier. A signer interacts with Notation CLI or builds application with Notation library to sign artifacts. A verifier interacts with Notation CLI or builds application with Notation library to verify artifacts against signatures.

Reviewer 4: ```suggestion
- **User**: The act could be a signer or a verifier. A signer interacts with Notation CLI or builds application using Notation library to sign artifacts. A verifier interacts with Notation CLI or builds application using Notation library to verify artifacts against signatures.
```

> - **User**: The User could be a signer or a verifier. A signer interacts with Notation CLI or builds application with Notation library to sign artifacts. A verifier interacts with Notation CLI or builds application with Notation library to verify artifacts against signatures.

Reviewer 4: IMO we should remove `builds application with Notation library` from this doc and only talk about Notation CLI because application can decide not to store trust policy or trust-store on disk which changes the threat model. Also, this will make doc more readable.

> - **Artifact Builder**: The actor who produces software artifacts in a build system. The software artifacts to be signed are arbitrary artifacts, which could be container images, helm charts in the context of Cloud, SBOM (Software Bill of Materials) in the context of secure supply chain, and other artifacts. Normally the artifacts are produced by the build process with vulnerabilities scanned, analyzed and fixed. Notation supports signing and verifying artifacts stored in registries or in the file system as OCI image layout.

Reviewer 4: nit: 
```suggestion
- **Artifact Builder**: The actor who is responsible for producing software artifacts using build systems. Notation supports signing and verification of software artifacts including but are not limited to container images, helm charts, SBOM (Software Bill of Materials). These artifacts can be stored either in remote registries or locally on disk using [OCI image layout](https://github.com/opencontainers/image-spec/blob/v1.0.0/image-layout.md).
```

  - Artifact in registries: Take container images as an example, users can use `docker` commands to produce container images, and store them to a registry. A container image in a registry can be referenced by a fully qualified path, for example `registry.wabbit-networks.io/net-monitor@sha256:xxx`. SBOM is another example, which are generated by tools during build process, stored as OCI artifacts in OCI-compliant registries, and referenced by a fully qualified path as well as container images.

>   - Artifact as OCI image layout: OCI image layout is a directory structure that contains files and folders that refer to an OCI image. See [OCI image layout spec](https://github.com/opencontainers/image-spec/blob/v1.0.0/image-layout.md) for details.

Reviewer 4: IMO we can remove this as they are not adding much value.

Author: It provides an option to sign artifacts before storing them in the registry first.

- **File System**: The file system where Notation or application built with Notation library is installed. For Notation signing artifacts as OCI image layout, it is the place where artifacts and associated signatures are stored. File system is within Notation trust boundary. Users or organizations should follow the security best practices to ensure the file system is safe to be used before proceeding any Notation commands. There are different directories and files as follows:
  - Notation config file, where general configuration is stored
  - Notation trust policy file, where trust policy configuration is stored
  - Notation trust store, where certificate files are stored to validate signatures
  - Notation plugin directory, where Notation plugins are installed

>   - Credential store, where the registry or Key management credentials are stored

Reviewer 4: I would replace FileSystem with Host/ExecutionEnvironment because user should even secure process not just file system.
```suggestion
- **Host/Execution environment**: The execution environment includes the host and file system where notation CLI will be deployed/installed and executed. Notation works with a shared responsibility model which means users/organizations are responsible for securing the notation execution environment. Some of the security best practices are described [here](Add-link-to-our doc page).  Following are the various directories and files used by the Notation:
  - `config.json` file is used to store various configurations such as credential store information, etc
  - `trustpolicy.json` file is used to store trust policy-related data.
  - `truststore` directory is used to store various trusted certificates used for verifying the signature.
  - `localkeys` directory is used to store test keys.
  -  `plugins` directory is used to store various plugins.
```
Notation uses credential stores to securely store the registry credentials.

Author: Accepted most comments except "Some of the security best practices are described [here](Add-link-to-our doc page). ", which will be added later once the document is available.

> - **Notation Plugin**: An external component built by different vendors, and integrated to Notation as a signing plugin for signature generation, or as a verifying plugin for signature verification, see [plugin spec](https://github.com/notaryproject/notaryproject/blob/v1.0.0-rc.2/specs/plugin-extensibility.md) for details.

Reviewer 4: ```suggestion
- **Notation Plugin**: An external component/binary that can integrate as one of the steps in Notation’s workflow for signature generation or verification, see [plugin spec](https://github.com/notaryproject/notaryproject/blob/v1.0.0-rc.2/specs/plugin-extensibility.md) for details.
```

> - **Registry**: A registry for Docker and Open Container Initiative (OCI) images, with support for all OCI artifacts. It is the storage where artifacts are pushed to by artifact builder. Registries are outside Notation trust boundary.

Reviewer 4: ```suggestion
- **Registry**: A registry stores images as well as related content formats, such as Helm charts, OCI artifacts  built to the OCI image format specification.
```

Author: I would like to keep the last sentence, which is "Registries are outside Notation trust boundary"

> - **KMS**: A remote Key Management Service provides key and associated certificate management, and remote signing service. KMSs are outside Notation trust boundary.

Reviewer 4: ```suggestion
- **KMS**: A Key Management System or Signing Service which is responsible for generation and management of signing materials such as certificate along with private key. KMS or Signing Service is outside Notation trust boundary.
```

> - **OCSP Responder**: The [Online Certificate Status Protocol (OCSP)](https://en.wikipedia.org/wiki/Online_Certificate_Status_Protocol) is an Internet protocol used for obtaining the revocation status of an X.509 digital certificate. Notation uses OCSP to validate certificate revocation status from OCSP Responder.

Reviewer 5: Reference RFC doc (i.e. [RFC 2560](https://www.rfc-editor.org/rfc/rfc2560)) instead of Wikipedia since Wikipedia cannot be used as an academic knowledge source.

> - **OCSP Responder**: The [Online Certificate Status Protocol (OCSP)](https://www.rfc-editor.org/rfc/rfc6960) is an Internet protocol used for obtaining the revocation status of an X.509 digital certificate. Notation uses OCSP to validate certificate revocation status from OCSP Responder.

Reviewer 4: ```suggestion
- **OCSP Responder**: The [Online Certificate Status Protocol (OCSP)](https://www.rfc-editor.org/rfc/rfc6960) is an Internet protocol used for obtaining the revocation status of an X.509 digital certificate. Notation uses OCSP to check the certificate's revocation status.
```

## Notation sign artifacts using remote key

### Data flow

A signer uses Notation to sign artifacts with the private key stored in a KMS, and store signatures in the file system or in a registry.

The following diagram illustrates Notation signing artifacts stored in registries with keys stored in KMS:

> ![image](../media/notation-sign-remote.svg)

Reviewer 5: "Step 3 login" is not a valid step as `login` is a pure client-side concept. Recommend to remove it.

Reviewer 5: If you do want to focus on `login`. It will be
1. Interact
2. **Ping registry**
3. **Write registry credentials** to credential store
4. Read registry credentials
5. Fetch descriptor of the artifact
6. ...

where steps 2 & 3 together are called **Login**.

> A signer uses Notation to sign artifacts with the private key stored in a KMS, and store signatures in the file system or in a registry.

Reviewer 4: ```suggestion
A signer uses Notation to sign the artifacts  and store signatures either locally on the disk or in a registry.
```

The following diagram illustrates Notation signing artifacts as OCI image layout with keys stored in KMS:

![image](../media/notation-sign-local.svg)

> According to [Notary signature specification](../specs/signature-specification.md), the signing payload is a JSON document which requires the [OCI descriptor](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/descriptor.md) of the artifact. For artifacts stored in a registry, the descriptor info can be retrieved by inspecting the [image manifest](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md) of the artifact. For artifacts stored as an [OCI Image layout](https://github.com/opencontainers/image-spec/blob/v1.0.0/image-layout.md), the descriptor info can be retrieved from the `index.json` file from OCI Image layout.

Reviewer 5: Just a question: do you want to link the latest version of image-spec?

Author: No. I want to use stable version for image-layout and descriptor. I will update the version of descriptor.md to v1.0.0 as well.

> According to [Notary signature specification](../specs/signature-specification.md), the signing payload is a JSON document which requires the [OCI descriptor](https://github.com/opencontainers/image-spec/blob/v1.0.0/descriptor.md) of the artifact. For artifacts stored in a registry, the descriptor info can be retrieved by inspecting the [image manifest](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md) of the artifact. For artifacts stored as an [OCI Image layout](https://github.com/opencontainers/image-spec/blob/v1.0.0/image-layout.md), the descriptor info can be retrieved from the `index.json` file from OCI Image layout.

Reviewer 4: ```suggestion
As per [Notary signature specification](../specs/signature-specification.md), the signature payload is the [OCI descriptor](https://github.com/opencontainers/image-spec/blob/v1.0.0/descriptor.md) of the target artifact. For artifacts stored in a registry, the descriptor can be retrieved from [image manifest](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md) stored in the registry. For artifacts stored as an [OCI Image layout](https://github.com/opencontainers/image-spec/blob/v1.0.0/image-layout.md), the descriptor can be retrieved by inspecting the `index.json` file of OCI Image layout.
```

### Threats and Mitigation

|   Title                      | STRIDE threat type     | Threat Status | Priority | Entity          | Description   | Mitigation   |
| ---------------------------- | ---------------------- | ------------- | -------- | --------------- | ------------- | ------------ |

> | Compromised signing key      | Information disclosure | Mitigated     | High     | Signer          | Signing keys are compromised, and this may lead to arbitrary artifacts being signed by attackers | Rotate the signing key, and revoke the associated certificates |

Reviewer 4: The mitigation would be to store the keys in hardware security module (HSM) and rotate them periodically. In case a signing material is compromised, user must revoke them.

| Inaccessible Registry        | Denial of Service      | Mitigated     | High     | Registry        | Registry is not able to service incoming requests or perform up to spec, thus users are unable to publish any artifacts | Notation signs artifacts stored in local file system and stored signatures locally |

> | Serving tampered artifacts   | Tampering              | Mitigated     | High     | Registry        | Data stored in registries are tampered, this may lead to malicious artifacts being signed | Notation signs artifacts stored in local file system, and stores signatures locally or in the registry |

Reviewer 4: If a registry is compromised and returns unexpected data, the mitigation here is that the signature payload includes the digest of the artifact and notation verified that the digest specified by user and the digest returned by notation are same.

Reviewer 4: Also this would fall under Spoofing,

Author: Registry is not with the trust boundary of Notation, so the artifacts stored in the registry could be already tampered with. Then signing and verifying don't help. The mitigation is to sign artifacts on the filesystem (Filesystem is within trust boundary.) before pushing the artifacts into registry.

> | Disclosure of signature info | Information disclosure | Mitigated     | High     | Registry        | Attackers eavesdrop on registries, this may lead to the disclosure of sensitive signature info | Users should interact with trusted registries via TLS or users can sign artifacts stored in local file system and store signatures locally |

Reviewer 3: Are the mitigations supposed to be how a user is supposed to respond to potential threats or how notation mitigates potential threats? My assumption is that it should be the latter, but most mitigations seem to be the former.

Specifically for this threat/mitigation, rather than "Users should interact with trusted registries via TLS", should the mitigation be framed as "Notation interacts with registries via TLS by default"?

For other threats (like "Compromised Signing Key"), is there nothing that Notation itself does to prevent this scenario? The mitigation makes it seem like there's nothing the user can do except rotate and re-sign artifacts without conveying how Notation is secure.

Author: @byronchien 

> Specifically for this threat/mitigation, rather than "Users should interact with trusted registries via TLS", should the mitigation be framed as "Notation interacts with registries via TLS by default"?

This is better, I will change it and try to update other mitigations

>For other threats (like "Compromised Signing Key"), is there nothing that Notation itself does to prevent this scenario? The mitigation makes it seem like there's nothing the user can do except rotate and re-sign artifacts without conveying how Notation is secure.

For this specific threat, there is no other way except rotating and re-sign artifacts since key is compromised.

Reviewer 3: for threats that hinge on the underlying filesystem, can we call out that the user is responsible for making sure that the filesystem is secured? there might be nothing that the user can do once it happens, but there they can take action to help prevent the threat from happening.

Author: @byronchien Please check the latest update.

> | Disclosure of signature info | Information disclosure | Mitigated     | High     | Registry        | Attackers eavesdrop on registries, this may lead to the disclosure of sensitive signature info | Notation interacts with trusted registries via TLS or signs artifacts stored in local file system, and then stores signatures locally |

Reviewer 4: I would uplevel this to call MitM attack. Also storing signatures locally is not the mitigation.

Author: The [STRIDE model](https://owasp.org/www-community/Threat_Modeling_Process#stride-threat-list) is used for the threat modeling, so there is no MitM attack type defined by this model.  Registry is not within the trust boundary of Notation, so signing and storing artifacts and signatures locally can be the mitigation. Could you share the reason that storing signatures locally is not the mitigation?

> | Tampered files               | Tampering              | Mitigated     | High     | File System     | Files read by Notation are tampered, this may lead to failure of notation sign operation | Notation is installed in trusted infrastructure with right file permissions |

Reviewer 4: ```suggestion
| Tampered files               | Tampering              | Mitigated     | High     | File System     | Files read by Notation are tampered, this may lead to failure of notation sign operation | Notation must be installed on secure and trusted infrastructure(shared responsibility mode) |
```

| Disclosure of files          | Information disclosure | Mitigated     | High     | File System     | Files are accessed by attackers, this may lead to disclosure of notation config files and credential files | Notation is installed in trusted infrastructure with right file permissions. Notation stores the credentials securely using credential store. Notation interacts with notation plugin to access signing keys via remote KMS or a HSM device |

> | Malicious plugin             | Tampering              | Mitigated     | High     | Notation Plugin | A malicious plugin is installed, this may lead to arbitrary code being executed by attackers | Verify plugin package before installation or plugin is installed by authorized users |

Reviewer 4: ```suggestion
| Malicious plugin             | Tampering              | Mitigated     | High     | Notation Plugin | A malicious plugin is installed, this may lead to arbitrary code being executed by attackers | Verify the integrity and authenticity of plugin before using plugin
```

## Notation verify artifacts against signatures

### Data flow

A verifier uses Notation to verify artifacts against signatures stored in the file system or in a registry to ensure the authenticity and integrity of the artifacts before deploying the artifacts.

The following diagram illustrates Notation verifying artifacts stored in a registry:

> ![image](../media/notation-verify-remote.svg)

Reviewer 5: The arrow to the OCSP responder seems has wrong direction.

The following diagram illustrates Notation verifying artifacts stored in a registry with a verification plugin:

> ![image](../media/notation-verify-plugin.svg)

Reviewer 5: The arrow to the OCSP responder seems has wrong direction.

The following diagram illustrates Notation verifying artifacts stored as OCI image layout in the file system:

> ![image](../media/notation-verify-local.svg)

Reviewer 5: Missing OCSP responder

> ![Notation verifies artifacts stored in registries](../media/notation-verify-remote.svg)

Reviewer 4: nit: We call revocation as last item in verification.

Author: Sorry, I didn't get it. Could you clarify your comment?

> ![Notation verifies artifacts as OCI image layout in the filesystem](../media/notation-verify-local.svg)

Reviewer 5: Missing trust boundary for OCSP responder.

The certificates trusted by the verifier are stored in Notation trust store in the file system. Users can download the certificates from KMS or request CA vendor for CA certificates.

### Threats and Mitigation

|   Title                      | STRIDE threat type     | Threat Status | Priority | Entity          | Description   | Mitigation   |
| ---------------------------- | ---------------------- | ------------- | -------- | --------------- | ------------- | ------------ |

> | Inaccessible Registry        | Denial of Service      | Mitigated     | High     | Registry        | Registry is not able to service incoming requests or perform up to spec, thus users are unable to verify artifacts | Notation verifies signatures stored locally |

Reviewer 4: ```suggestion
| Inaccessible Registry        | Denial of Service      | Mitigated     | High     | Registry        | Registry is not able to service incoming requests or perform up to spec, thus users are unable to verify artifacts | Notation support verification of signatures stored locally |
```

> | Serving tampered artifacts   | Tampering              | Mitigated     | High     | Registry        | Data stored in registries are tampered, this may lead to malicious artifacts being consumed | Notation verifies artifacts before artifacts are consumed |

Reviewer 4: ```suggestion
| Serving tampered artifacts   | Tampering              | Mitigated     | High     | Registry        | Data stored in registries are tampered, this may lead to malicious artifacts being consumed | Notation verifies that the  that the digest specified by user and the digest store in signature are same. Notation does support verification using tag but its highly discouraged by emitting warning.
```

> | Disclosure of signature info | Information disclosure | Mitigated     | High     | Registry        | Attackers eavesdrop on registries, this may lead to the disclosure of sensitive signature info | Users should interact with trusted registries via TLS or Users can sign artifacts stored in local file system and store signature locally |

Author: I agree with @Two-Hearts to remove this from verifying scenario. @shizhMSFT any comments?

> | Disclosure of signature info | Information disclosure | Mitigated     | High     | Registry        | Attackers eavesdrop on registries, this may lead to the disclosure of sensitive signature info | Users should interact with trusted registries via TLS or Users can sign artifacts stored in local file system and store signature locally |

Reviewer 5: Although artifacts cannot be tampered by the attackers as they are protected by the signatures, the artifacts and the signatures may still be confidential. Disclosure of signature info may result in privacy incident.

For instance, signer info might be leaked from a signature when it is transferred via plain HTTP. If the signer info indicates a product team of a company, which is never announced, the adversary may gain knowledge that that company is preparing for a new product release. Then the adversary can manage to announce its own product first to capture more market.

> | Disclosure of signature info | Information disclosure | Mitigated     | High     | Registry        | Attackers eavesdrop on registries, this may lead to the disclosure of sensitive signature info | Users should interact with trusted registries via TLS or Users can sign artifacts stored in local file system and store signature locally |

Author: Thanks @shizhMSFT for explanation. This comment was resolved then.

> | Disclosure of signature info | Information disclosure | Mitigated     | High     | Registry        | Attackers eavesdrop on registries, this may lead to the disclosure of sensitive signature info | Notation interacts with trusted registries via TLS by default or Notation signs artifacts stored in local file system, and then stores signatures locally |

Reviewer 4: same as signing comment

Author: Explained in previous comment

| Tampered files               | Tampering              | Mitigated     | High     | File System      | Files read by Notation are tampered, this may lead to failure of notation verify operation or verification bypassed | Use trusted infrastructure with right file permissions |
| Disclosure of files          | Information disclosure | Mitigated     | High     | File System      | Files are accessed by attackers, this may lead to disclosure of notation config files and credential files | Use trusted infrastructure with right file permissions. Users can use credential store to securely store the credentials |
| Malicious plugin             | Tampering              | Mitigated     | High     | Notation Plugin | A malicious plugin is installed, this may lead to arbitrary code being executed by attackers | Verify plugin package before installation or plugin is installed by authorized users |

> | Inaccessible OCSP Responder  | Denial of Service      | Mitigated     | High     | OCSP Responder  | OCSP Responder is not able to service incoming requests or perform up to spec, thus users are unable to validate certificate revocation status | Users can configure trust policy to skip certificate revocation status check |

Author: Revocation checking with CRL was not supported by Notation v1 release, so your proposed mitigation doesn't work. Any idea from @shizhMSFT and @priteshbandi?

> | Inaccessible OCSP Responder  | Denial of Service      | Mitigated     | High     | OCSP Responder  | OCSP Responder is not able to service incoming requests or perform up to spec, thus users are unable to validate certificate revocation status | Users can configure trust policy to skip certificate revocation status check |

Reviewer 4: If OCSP Responder inaccessible then Notation cannot mitigate the threat (I would say its unmitigated) because it’s not in notation purview to control OCSP behaviour and verification should fail if revocation is configured as enforce and OCSP is inaccessible.
The best user can do is to change revocation level to warn to bypass the issue but that again is not a mitigation.

> | Inaccessible OCSP Responder  | Denial of Service      | Mitigated     | High     | OCSP Responder  | OCSP Responder is not able to service incoming requests or perform up to spec, thus users are unable to validate certificate revocation status | Users can configure trust policy to skip certificate revocation status check |

Reviewer 5: You may consider OSCP is a more efficient replacement of CRL but they all require network access.

I agree with @priteshbandi that we move revocation check out of the critical path to avoid availability issues.

# Notation Threat Model

## Overview

Notation aims for signing and validating software artifacts, ensuring they have not been tampered with and providing security policies to determine which validated artifacts are allowed to be used in your systems.

The following diagram illustrates the architecture and components.

![image](../media/notation-system.png)

> - **User**: The User could be a signer or a verifier. The signer interacts with Notation CLI or builds application with Notation library to sign artifacts. The verifier interacts with Notation CLI or build application with Notation library to verify artifacts against signatures.

Reviewer 2: A user can be both a signer and a verifier.
```suggestion
- **User**: The User could be a signer and/or a verifier. A signer interacts with Notation CLI or builds application with Notation library to sign artifacts. A verifier interacts with Notation CLI or builds application with Notation library to verify artifacts against signatures.
```

> - **Artifact Builder**: The actor who produces software artifacts in a build system. The software artifacts to be signed are arbitrary artifacts, which could be container images, and helm charts in the context of Cloud, or SBOM (Software Bill of Materials) in the context of secure supply chain, or other artifacts. Normally the artifacts are produced by the build process with vulnerabilities scanned, analyzed and fixed. Notation support signing and verifying both local and remote artifacts.

Reviewer 2: ```suggestion
- **Artifact Builder**: The actor who produces software artifacts in a build system. The software artifacts to be signed are arbitrary artifacts, which could be container images, helm charts in the context of Cloud, SBOM (Software Bill of Materials) in the context of secure supply chain, and other artifacts. Normally the artifacts are produced by the build process with vulnerabilities scanned, analyzed and fixed. Notation supports signing and verifying both local and remote artifacts.
```

> - **Artifact Builder**: The actor who produces software artifacts in a build system. The software artifacts to be signed are arbitrary artifacts, which could be container images, and helm charts in the context of Cloud, or SBOM (Software Bill of Materials) in the context of secure supply chain, or other artifacts. Normally the artifacts are produced by the build process with vulnerabilities scanned, analyzed and fixed. Notation support signing and verifying both local and remote artifacts.

Reviewer 5: Avoid using the term `local` and `remote` as they have different meanings in general.

Usually, `local` means _on the same device_ while `remote` requires network interactions.

Reviewer 5: For instances,
- Artifact in the registry running on `127.0.0.1:5000` should be considered as **local**.
- Artifacts stored in a NAS but mount to the file system should be considered as **remote**.

Therefore, we need more accurate terms.

>   - Local artifact: The artifact is stored in a local storage as an arbitrary file, or [OCI Image Layout](https://github.com/opencontainers/image-spec/blob/v1.0.0/image-layout.md) or an [OCI descriptor](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/descriptor.md) file created for an artifact

Reviewer 2: ```suggestion
  - Local artifact: The artifact is stored in a local storage, such as an arbitrary file, [OCI Image Layout](https://github.com/opencontainers/image-spec/blob/v1.0.0/image-layout.md), or an [OCI descriptor](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/descriptor.md) file created for an artifact.
```

>   - Remote artifact: The artifact is stored in a remote registry. Take container images as an example, users can use `docker` commands to produce container images, and store them to a remote registry. The container images in a registry can be referenced by a fully qualified path, for example `registry.wabbit-networks.io\net-monitor@sha256:xxx`. Another example is about SBOM, which are generated by tools during build process, stored as OCI artifacts in OCI-compliant registries, and referenced by a fully qualified path as well as container images

Reviewer 2: ```suggestion
  - Remote artifact: The artifact is stored in a remote registry. Take container images as an example, users can use `docker` commands to produce container images and store them to a remote registry. A container image in a registry can be referenced by a fully qualified path, for example `registry.wabbit-networks.io\net-monitor@sha256:xxx`. SBOM is another example, which are generated by tools during build process, stored as OCI artifacts in OCI-compliant registries, and referenced by a fully qualified path as well as container images.
```

>   - Remote artifact: The artifact is stored in a remote registry. Take container images as an example, users can use `docker` commands to produce container images, and store them to a remote registry. The container images in a registry can be referenced by a fully qualified path, for example `registry.wabbit-networks.io\net-monitor@sha256:xxx`. Another example is about SBOM, which are generated by tools during build process, stored as OCI artifacts in OCI-compliant registries, and referenced by a fully qualified path as well as container images

Reviewer 5: use forward slash: `registry.wabbit-networks.io\net-monitor@sha256:xxx` -->  `registry.wabbit-networks.io/net-monitor@sha256:xxx`

- **File System**: The file system where Notation or application built with Notation library is installed. For Notation signing local artifact, it is the place where local artifact and associated signatures are stored. There are different stores as follows

>   - Notation config store, where Notation configuration file are stored, for example, trust policy configuration

Reviewer 2: As mentioned above, looks like `Notation config store` is a new term. Do we have any other docs/specs defined/mentioned it?

  - Notation trust store, where certificates are added to validate signatures
  - Notation plugin store, where Notation plugins are installed
  - Credential store, where the registry or Key management credentials are stored
- **Notation Plugin**: An external component built by different vendors, and integrated to Notation as a signing plugin for signature generation, or as a verifying plugin for signature verification, see [plugin spec](https://github.com/notaryproject/notaryproject/blob/v1.0.0-rc.2/specs/plugin-extensibility.md) for details.

> - **Registry**: A remote registry of Docker and Open Container Initiative (OCI) images, with support for all OCI artifacts. It is the remote storage where artifacts are pushed to by artifact builder. Registries are outside Notation trust boundary.

Reviewer 2: ```suggestion
- **Registry**: A remote registry for Docker and Open Container Initiative (OCI) images, with support for all OCI artifacts. It is the remote storage where artifacts are pushed to by artifact builder. Registries are outside of Notation's trust boundary.
```

> - **KMS**: A remote Key Management Service provides key and associated certificate management, and remote signing service. KMSs are outside Notation trust boundary.

Reviewer 2: ```suggestion
- **KMS**: A remote Key Management Service provides key and associated certificate management, and remote signing service. KMSs are outside of Notation's trust boundary.
```

Author: `outside` can be used as prep, so I will stick to current expression.

## Notation sign artifacts using remote key

### Data flow

> A signer uses Notation to sign artifacts with the private key stored in a KMS, and store signatures in a local file system or in a remote registry.

Reviewer 6: When data flows intersect a particular boundary (network, client machine, domain, etc.), we may need to specify and draw the trust boundary in the data flow diagram.

Author: I will modify the description of the dot line since it is originally used for marking the trust boundary. I will also try to add text accordingly

The following diagram illustrates Notation signing remote artifacts with KMS:

![image](../media/notation-sign-remote.png)

The following diagram illustrates Notation signing local artifacts with KMS:

![image](../media/notation-sign-local.png)

> According to [Notary signature specification](../specs/signature-specification.md), the signing payload is a JSON document which requires the [OCI descriptor](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/descriptor.md) of the artifact. For artifacts stored in a registry, the descriptor info can be retrieved by inspecting the [image manifest](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md) of the artifact. For artifacts stored locally as an [OCI Image layout](https://github.com/opencontainers/image-spec/blob/v1.0.0/image-layout.md), the descriptor info can be retrieved from the `index.json` file from OCI Image layout. Users can also pass descriptor info as a JSON file to Notation directly.

Reviewer 2: `Users can also pass descriptor info as a JSON file to Notation directly.` The corresponding implementation is not in Notation as of today. Should we include this in our Notation 1.0 scope?

Author: It's not in 1.0 scope. I think I should remove it for now.

### Threats and Mitigation

|   Title                      | STRIDE threat type     | Threat Status | Priority | Entity          | Description   | Mitigation   |
| ---------------------------- | ---------------------- | ------------- | -------- | --------------- | ------------- | ------------ |
| Compromised signing key      | Information disclosure | Mitigated     | High     | Signer          | Signing keys are compromised, and this may lead to arbitrary artifacts being signed by attackers | Rotate the signing key, and revoke the associated certificates |

> | Inaccessible Registry        | Denial of Service      | Mitigated     | High     | Registry        | Registry is not able to service incoming requests or perform up to spec, thus users are unable to publish any artifacts | Users sign offline artifacts and publish via other medias |

Reviewer 6: Can we revise the Mitigation "Users sign offline artifacts and publish via other medias" to "Users sign local artifacts offline and publish via other medias", which aligns with the term "local artifacts" we used above?

| Tampered files               | Tampering              | Mitigated     | High     | File System     | Files read by Notation are tampered, this may lead to failure of notation sign operation | Use trusted infrastructure with right file permissions |
| Disclosure of files          | Information disclosure | Mitigated     | High     | File System     | Files are accessed by attackers, this may lead to disclosure of notation config files and credential files | Use trusted infrastructure with right file permissions. Users can use credential store to securely store the credentials. Users can use notation plugin to access signing keys via remote KMS or a HSM device |
| Malicious plugin             | Tampering              | Mitigated     | High     | Notation Plugin | A malicious plugin is installed, this may lead to arbitrary code being executed by attackers | Verify plugin package before installation or plugin is installed by authorized users |

## Notation verify artifacts against signatures

### Data flow

> A verifier uses Notation to verify artifacts against signatures stored locally or in a remote registry to ensure the authenticity and integrity of the artifacts before using the artifacts.

Reviewer 2: ```suggestion
A verifier uses Notation to verify artifacts against signatures stored locally or in a remote registry to ensure the authenticity and integrity of the artifacts before deploying the artifacts.
```

The following diagram illustrates Notation verifying remote artifacts:

![image](../media/notation-verify-remote.png)

The following diagram illustrates Notation verifying remote artifacts with a verification plugin:

![image](../media/notation-verify-plugin.png)

The following diagram illustrates Notation verifying local artifacts:

![image](../media/notation-verify-local.png)

> The certificates used for signature verification are stored in Notation trust store in the file system. Users can download the certificates from KMS or request CA vendor for CA certificates.

Reviewer 2: ```suggestion
The certificates trusted by the verifier are stored in Notation trust store in the file system. Users can download the certificates from KMS or request CA vendor for CA certificates.
```
